### PR TITLE
Attributes Option and Directory Callback

### PR DESCRIPTION
**Pull request by edave**
Adds the ability to have additional attributes from FS.Stats included and also included a callback function for on Directories.

### CODE REVIEW

File: lib/directory-tree.js [hunk 1]
```diff
@@ -30,32 +30,47 @@ function directoryTree (path, options, onEachFile) {
...
+			options.attributes.forEach((attribute) => {
+				item[attribute] = stats[attribute];
+			});
+		}
 		item.children = dirData
 			.map(child => directoryTree(PATH.join(path, child), options, onEachFile))
 			.filter(e => !!e);
 		item.size = item.children.reduce((prev, cur) => prev + cur.size, 0);
 		item.type = constants.DIRECTORY;
+		if (onEachDirectory) {
```
Comment (mihneadb) [inline lib/directory-tree.js]: What if we keep one single callback and the user has to check the item's type inside? Since you'll get the stats now.

Comment (edave) [reply lib/directory-tree.js]: I like that even better. Will push up a commit with that soon

**Comment by edave** [comment]
@mihneadb I'm not sure why the TravisCI build failed, it must be something to do with accessing the file time stamps in the TravisCI container. All the tests pass on MacOS and the failing test is just comparing what `FS.stats` returns for the file's created & modified timestamp in milliseconds). I'll try a different approach, but it may involve me pushing a few commits to test TravisCI.

**Comment by edave** [comment]
@mihneadb I remember now why I didn't extend the file callback to run on directories as well- it leads to odd (and probably unexpected) behavior when you specify to only return files with a certain extension because now you're getting callbacks for those files _and_ the directories. We could go with just a single callback, but I suspect it would break how many people are using the file callback.

**Comment by mihneadb** [comment]
Gotcha. OK, let's go with that. Can you pls update the readme as well to illustrate that? Thanks!

**Comment by edave** [comment]
@mihneadb Yep, the README has instructions for that as well- let me know if they're not clear.

**Comment by mihneadb** [comment]
@edave sorry, missed it somehow through the Diff view. Thanks! Can you please resolve the conflicts?

**Comment by rizsc** [comment]
@mihneadb @edave thanks for this awesome feature, when this will be released? I really need directory callback to modify full path, thanks!

**Comment by mihneadb** [comment]
This branch has conflicts with master so it can't be merged right now. @rizsc you can walk the resulting tree and change the paths in place, meanwhile.

**Comment by edave** [comment]
@mihneadb apologies, missed the conflicts. Will resolve later this week!

**Comment by mihneadb** [comment]
@edave somehow it still displays commits. I think you need to fetch the latest upstream master first.

**Comment by rmdwirizki** [comment]
Hope this feature will be released soon! I need stats object to get the inode number for every file or directory to generate some unique id.

**Comment by mihneadb** [comment]
Pinging @edave :).

**Pull request merged**